File: c/list-ops/list_ops.c
#include "list_ops.h"
#include <memory.h>


list_t *new_list(size_t length, list_element_t elements[]) {
    list_t *dup = malloc(sizeof(list_t) + sizeof(list_element_t) * length);
    dup->length = length;
    memcpy(dup->elements, elements, length * sizeof(list_element_t));
    return dup;
}

list_t *append_list(list_t *list1, list_t *list2) {
    list_t *dup = malloc(
            sizeof(list_t) + sizeof(list_element_t) * list1->length + sizeof(list_element_t) * list2->length);
    dup->length = list1->length + list2->length;
    memcpy(dup->elements, list1->elements, sizeof(list_element_t) * list1->length);
    memcpy(dup->elements + list1->length, list2->elements, sizeof(list_element_t) * list2->length);
    return dup;
}

void delete_list(list_t *list) {
    free(list);
}

list_t *filter_list(list_t *list, bool(*filter)(list_element_t)) {
    list_element_t *filtered_List = malloc(sizeof(list_t) + sizeof(list_element_t) * list->length);
    int j = 0;
    for (size_t i = 0; i < list->length; ++i) {
        if (filter(list->elements[i])) {
            filtered_List[j++] = list->elements[i];
        }
    }
    return new_list(j, filtered_List);
}

size_t length_list(list_t *list) {
    return list->length;
}

list_t *map_list(list_t *list, list_element_t(*map)(list_element_t)) {
    list_element_t *mapped_List = malloc(sizeof(list_t) + sizeof(list_element_t) * list->length);
    int j = 0;
    for (size_t i = 0; i < list->length; ++i) {
        mapped_List[j++] = map(list->elements[i]);
    }
    return new_list(j, mapped_List);
}

list_element_t foldl_list(list_t *list, list_element_t initial, list_element_t(*foldl)(list_element_t, list_element_t)) {
    list_element_t acc = initial;
    for (size_t i = 0; i < list->length; ++i) {
        acc = foldl(list->elements[i], acc);
    }
    return acc;
}

list_element_t foldr_list(list_t *list, list_element_t initial, list_element_t(*foldr)(list_element_t, list_element_t)) {
    list_element_t acc = initial;
    for (int i = (int) list->length - 1; i >= 0; --i) {
        acc = foldr(list->elements[i], acc);
    }
    return acc;
}

list_t *reverse_list(list_t *list) {
    list_t *rvs = malloc(sizeof(list_t) + sizeof(list_element_t) * list->length);
    rvs->length = list->length;
    int j = 0;
    for (int i = (int) list->length - 1; i >= 0; --i) {
        rvs->elements[i] = list->elements[j++];
    }
    return rvs;
}

File: go/list-ops/list_ops.go
package listops

type binFunc func(int, int) int
type predFunc func(int) bool
type unaryFunc func(int) int
type IntList []int

func (l IntList) Foldr(f binFunc, initial int) int {
	for i := l.Length() - 1; i >= 0; i-- {
		initial = f(l[i], initial)
	}
	return initial
}
func (l IntList) Foldl(f binFunc, initial int) int {
	for _, i := range l {
		initial = f(initial, i)
	}
	return initial
}
func (l IntList) Filter(f predFunc) IntList {
	r := IntList{}
	for _, i := range l {
		if f(i) {
			r = r.Append(IntList{i})
		}
	}
	return r
}
func (l IntList) Length() int {
	return l.Foldl(func(x, y int) int { return x + 1 }, 0)
}
func (l IntList) Map(f unaryFunc) IntList {
	r := IntList{}
	for _, i := range l {
		r = r.Append(IntList{f(i)})
	}
	return r
}
func (l IntList) Reverse() IntList {
	r := IntList{}
	for i := l.Length() - 1; i >= 0; i-- {
		r = r.Append(IntList{l[i]})
	}
	return r
}
func (l IntList) Append(lst IntList) IntList {
	r := make(IntList, len(l)+len(lst))
	i := 0
	for ; i < len(l); i++ {
		r[i] = l[i]
	}
	for j := 0; j < len(lst); j++ {
		r[i] = lst[j]
		i++
	}
	return r
}
func (l IntList) Concat(lsts []IntList) IntList {
	r := IntList{}.Append(l)
	for _, lst := range lsts {
		r = r.Append(lst)
	}
	return r
}
